### PR TITLE
Fixes for compile tests

### PR DESCRIPTION
**Pull request by mborland**
Fixes for failures from -Wzero-length-array (see #614 discussion) and other warnings.

### CODE REVIEW

**Comment by NAThompson** [comment]
All the diffs I understand I approve of!

**Comment by mborland** [comment]
This should be good to go. CI is clean

File: include/boost/math/special_functions/rsqrt.hpp [hunk 1]
```diff
@@ -15,7 +15,7 @@ template<typename Real>
 inline Real rsqrt(Real const & x)
 {
     using std::sqrt;
-    if constexpr (std::is_same_v<Real, float> || std::is_same_v<Real, double> || std::is_same_v<Real, long double>)
+    if constexpr (std::is_floating_point_v<Real>)
```
Comment (jzmaddock) [inline include/boost/math/special_functions/rsqrt.hpp:18]: @NAThompson : was the original intention here to exclude `__float128`?  If so replacing with is_floating_point *might* fail in -std=gnu++ mode.   Just double checking...

File: include/boost/math/special_functions/nonfinite_num_facets.hpp [hunk 2]
```diff
@@ -414,7 +414,7 @@ namespace boost {
         switch(peek_char(it, end, ct)) {
         case 'q':
         case 's':
-          if(flags_ && legacy)
+          if(flags_)
```
Comment (jzmaddock) [inline include/boost/math/special_functions/nonfinite_num_facets.hpp:417]: Looking at the rest of the code, I assume this should be `flags_&legacy` ?  @pabristow ?

Comment (pabristow) [reply include/boost/math/special_functions/nonfinite_num_facets.hpp:417]: I deny writing it! Looks reasonable, though the comments seem not very specific to me.  I can't see any reason to restrict to only legacy here. And the tests are thin.  And testing is difficult on account of the too many possible floating-point types and platforms.

Lesson to Standards Writers - make up your mind and don't duck out by making stuff implementation-dependent!

File: include/boost/math/special_functions/rsqrt.hpp (hunk 1 above)
Comment (NAThompson) [reply include/boost/math/special_functions/rsqrt.hpp:18]: Amazingly enough I remember this! Yes, that was the goal.

As you can see in the `else` branch, the `rsqrt` iteration starts with the `long double` approximation and then refines it with the Newton method for `1/sqrt(x)`. I verified that it is slower to call `1/sqrt(x)` for `float128` than to use the `long double` approximation and Newton iterate.

**Comment by mborland** [comment]
The one erroneous diff has been reverted so should be good to go pending further review.

**Pull request merged**